### PR TITLE
Master previous lunch order o2m psa

### PR DESCRIPTION
**Pull request by msa-odoo**
Add lunch widget to display a previous order.

Task: https://www.odoo.com/web#id=10450&view_type=form&model=project.task&menu_id=3942&action=327

Pad: https://pad.odoo.com/p/r.0Gr6bmJOGr2cAnGr

### CODE REVIEW

File: addons/lunch/static/src/js/lunch.js [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+            var self = this;
+            new openerp.web.Model("lunch.order.line").call('read', [parseInt(pref_id), ['product_id', 'supplier', 'note', 'price']])
+                .then(function(res){
+                    if (res.id == parseInt(pref_id)) {
+                        if(res.note == 'false') { res.note = ''; }
+                        records.push([0, 0, {'product_id': res.product_id[0], 'note': res.note, 'supplier': res.supplier[0], 'price': res.price}]);
+                        self.field_manager.set_values({order_line_ids: records});
+                    }
+            });
+        },
```
Comment (vme-odoo) [inline addons/lunch/static/src/js/lunch.js]: we can avoid this rpc call by storing data of get_previous_lunch_order_details in widget.

File: addons/lunch/static/src/js/lunch.js [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+        render_value: function() {
+            var self = this;
+            new openerp.web.Model("lunch.order.line").call('read',[self.get_value(),['product_id', 'note', 'price', 'category_id','currency_id']])
+            .then(function(data) {
+                if (_.isEmpty(data)) {
+                    self.$el.html(QWeb.render("lunch.no_preference_ids"));
+                } else {
+                    var categories = {};
+                    var currency = {};
+                    for (var i=0; i<data.length; i++) {
```
Comment (vme-odoo) [inline addons/lunch/static/src/js/lunch.js]: use _.each

File: addons/lunch/static/src/js/lunch.js [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+                        }else{
+                            categories[ data[i]['category_id'][1] ][ data[i]['product_id'][1] ]  = {};
+                            categories[ data[i]['category_id'][1] ][ data[i]['product_id'][1] ][ data[i].note ]  = detail;
+                        }
+                    }
+                    self.$el.html(QWeb.render("lunch.PreviousOrderLine", {'categories': categories, 'currency': currency}));
+                    self.$el.find('span.add_button').click( function(event) {
+                        var pref_id = $(event.currentTarget).attr('id');
+                        var records = self.field_manager.get_field_value('order_line_ids');
+                        self.update_lines(pref_id, records);
```
Comment (vme-odoo) [inline addons/lunch/static/src/js/lunch.js]: Why fetching order line secound time?

File: addons/lunch/static/src/xml/lunch.xml [hunk 4]
```diff
@@ -0,0 +1,44 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<templates>
+    <t t-name="lunch.PreviousOrder">
+        <div>
+            <div class="previous_order_lines">hello</div>
+        </div>
```
Comment (vme-odoo) [inline addons/lunch/static/src/xml/lunch.xml]: is it in spec or forgot to remove?

File: addons/lunch/lunch.py [hunk 5]
```diff
@@ -318,6 +177,7 @@ def fields_view_get(self, cr, uid, view_id=None, view_type=False, context=None,
         'date': fields.date.context_today,
         'state': 'new',
         'alerts': _default_alerts_get,
+        'previous_order_details': _default_get_lunch_order_detail,
```
Comment (vme-odoo) [inline addons/lunch/lunch.py]: why same private method *_default_get_lunch_order_detail* is call from defaults and also from  get_lunch_order_detail?

File: addons/lunch/lunch.py [hunk 6]
```diff
@@ -141,6 +119,21 @@ def _default_alerts_get(self, cr, uid, context=None):
...
 
+    def get_lunch_order_detail(self, cr, uid, ids, name, arg, context=None):
+        """
+        get the previous lunch order details to display on the order form
+        """
+        data = self._default_get_lunch_order_detail(cr, uid, context=None)
+        res = {}
+        for id in ids:
+            res[id] = data
+        return res
```
Comment (vme-odoo) [inline addons/lunch/lunch.py]: optimize code.

File: addons/lunch/static/src/js/lunch.js [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+                        if (!(data[i]['category_id'][1] in categories)){
+                            categories[ data[i]['category_id'][1] ] = {};
+                        }
+                        if (data[i]['product_id'][1] in categories[ data[i]['category_id'][1] ]) {
+                            categories[ data[i]['category_id'][1] ][ data[i]['product_id'][1] ][ data[i].note ]  = detail;
+                        }else{
+                            categories[ data[i]['category_id'][1] ][ data[i]['product_id'][1] ]  = {};
+                            categories[ data[i]['category_id'][1] ][ data[i]['product_id'][1] ][ data[i].note ]  = detail;
+                        }
+                    }
```
Comment (vme-odoo) [inline addons/lunch/static/src/js/lunch.js]: Looks too much static, if its compulsory it can be handle from qweb.

**Comment by vme-odoo** [comment]
some fixes and code optimize committed.

File: addons/lunch/lunch.py [hunk 8]
```diff
@@ -141,6 +119,18 @@ def _default_alerts_get(self, cr, uid, context=None):
...
 
+    def get_lunch_order_detail(self, cr, uid, ids, name, arg, context=None):
+        """
+        get the previous lunch order details to display on the order form
+        """
+        data = self._default_get_lunch_order_detail(cr, uid, context=None)
+        return { id: list(data) for id in ids }
+
+    def _default_get_lunch_order_detail(self, cr, uid, context=None):
+        cr.execute("select MAX(id) id from lunch_order_line where user_id=%s group by note,product_id order by id DESC;", [uid]);
```
Comment (tde-banana-odoo) [inline addons/lunch/lunch.py]: Probably
 - add a limit
 - execute -> standard query

File: addons/lunch/lunch.py [hunk 9]
```diff
@@ -311,6 +164,9 @@ def fields_view_get(self, cr, uid, view_id=None, view_type=False, context=None,
                                     ('partially','Partially Confirmed')] \
                                 ,'Status', readonly=True, select=True, copy=False),
         'alerts': fields.function(_alerts_get, string="Alerts", type='text'),
+        'previous_order_details': fields.function(get_lunch_order_detail, string="Previous Orders", type="one2many", relation="lunch.order.line"),
```
Comment (tde-banana-odoo) [inline addons/lunch/lunch.py]: Rename to previous_order_ids .

**Comment by jem-odoo** [comment]
Old PR, the new one is https://github.com/odoo/odoo/pull/5522

**Pull request closed without merging**